### PR TITLE
python3Packages.keyring-pass: init at 0.9.2

### LINKED ISSUE #241660: Package request: keyring-pass
**Project description**

Password Store (pass) backend for python's keyring 

**Metadata**

* homepage URL: https://pypi.org/project/keyring-pass/
* source URL: https://github.com/nazarewk/keyring_pass
* license: mit
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by nazarewk**
## Description of changes

fixes #241660 
implements https://github.com/nazarewk/keyring_pass/issues/15

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+, pythonOlder
+}:
+let
+  pname = "keyring-pass";
+  version = "0.8.1";
+  module_name = builtins.replaceStrings [ "-" ] [ "_" ] pname;
+in
+
+buildPythonPackage {
+  inherit pname version;
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:17]:
```suggestion
  pname = "keyring-pass";
  version = "0.8.1";
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+, keyring
+, pass
+, poetry-core
+, pythonOlder
+}:
+let
+  pname = "keyring-pass";
+  version = "0.8.1";
+  module_name = builtins.replaceStrings [ "-" ] [ "_" ] pname;
+in
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:14]:
```suggestion
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+buildPythonPackage {
+  inherit pname version;
+  disabled = pythonOlder "3.6";
+  format = "pyproject";
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:19]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+  src = fetchPypi {
+    # results in 404 without this replacement
+    pname = module_name;
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:23]:
```suggestion
    pname = "keyring_pass";
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    pass
+  ];
+
+  nativeCheckInputs = [
+    keyring
+    gnupg
+    pass
+  ];
+
+  doCheck = true;
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:42]:
```suggestion
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+  preCheck = ''
+    export PASSWORD_STORE_DIR="$TMPDIR/.password-store"
+    export GNUPGHOME="$TMPDIR/.gnupg"
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:46]:
```suggestion
    export HOME=$TMPDIR
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    runHook postCheck
+  '';
+
+  pythonImportsCheck = [
+    module_name
+  ];
+
+  meta = {
+    license = lib.licenses.mit;
+    description = "Password Store (pass) backend for python's keyring";
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:79]: Please order alphabetically.

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  nativeCheckInputs = [
+    keyring
+    gnupg
+    pass
+  ];
+
+  doCheck = true;
+
+  preCheck = ''
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:44]: just put this inside of `checkPhase`

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    Expire-Date: 1
+    EOF
+    # initialize password store
+    pass init test@example.com
+  '';
+
+  checkPhase = ''
+    runHook preCheck
+
+    # added `__init__` due to: No module named keyring_pass.__main__; 'keyring_pass' is a package and cannot be directly executed
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:67]: Does it help if we `rm -r keyring_pass` first and add `$out/...` to `$PYTHONPATH`?

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  version = "0.8.1";
+  module_name = builtins.replaceStrings [ "-" ] [ "_" ] pname;
+in
+
+buildPythonPackage {
+  inherit pname version;
+  disabled = pythonOlder "3.6";
+  format = "pyproject";
+
+  src = fetchPypi {
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:21]: I prefer `fetchFromGitHub` but it's up to you.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please fix the commit message.

File: pkgs/development/python-modules/keyring-pass/default.nix (hunk 2 above)
Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix:14]: there is still a matter of `version` used in 2 places

File: pkgs/development/python-modules/keyring-pass/default.nix (hunk 9 above)
Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix:67]: while doing so I noticed `$out` is already part of PYTHONPATH

File: pkgs/development/python-modules/keyring-pass/default.nix (hunk 10 above)
Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix:21]: I don't actually know how to build poetry package without `poetry2nix`, but otherwise wouldn't be opposed to `fetchFromGitHub`

Comment (dotlambda) [reply pkgs/development/python-modules/keyring-pass/default.nix:21]: Poetry-core should be enough. You can simply replace the `src`.

File: pkgs/development/python-modules/keyring-pass/default.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/development/python-modules/keyring-pass/default.nix]: You can take care of that using `rec`.

Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix]: isn't `rec` kind of hard to follow antipattern?

Comment (dotlambda) [reply pkgs/development/python-modules/keyring-pass/default.nix]: For Python packages there is no alternative yet.

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  inherit version;
+  disabled = pythonOlder "3.6";
+
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "nazarewk";
+    repo = "keyring_pass";
+    rev = "v${version}";
+    sha256 = hash;
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:25]:
```suggestion
    hash = "sha256-sTvJb67s78XzG0nOAsqIaJnMw7RFPpSVbGPH68sUJKE=";
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+, poetry-core
+, pythonOlder
+}:
+let
+  version = "0.8.1";
+  hash = "sha256-sTvJb67s78XzG0nOAsqIaJnMw7RFPpSVbGPH68sUJKE=";
+in
+buildPythonPackage {
+  pname = "keyring-pass";
+  inherit version;
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:16]:
```suggestion
  version = "0.8.1";
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+, keyring
+, pass
+, poetry-core
+, pythonOlder
+}:
+let
+  version = "0.8.1";
+  hash = "sha256-sTvJb67s78XzG0nOAsqIaJnMw7RFPpSVbGPH68sUJKE=";
+in
+buildPythonPackage {
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:14]:
```suggestion
buildPythonPackage rec {
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    rev = "v${version}";
+    sha256 = "sha256-Sf7eDOB3prH2s6BzdBtxewSweC0ibLXVxNHBJRRaJe4=";
+  };
+
+  nativeBuildInputs = [
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    pass
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:31]: This doesn't sound right cause it's not a Python dependency. Instead hardcode the path to the executable using a patch.

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  disabled = pythonOlder "3.6";
+
+  passthru.updateScript = nix-update-script { };
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "nazarewk";
+    repo = "keyring_pass";
+    rev = "v${version}";
+    sha256 = "sha256-Sf7eDOB3prH2s6BzdBtxewSweC0ibLXVxNHBJRRaJe4=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:23]:
```suggestion
    hadh = "sha256-Sf7eDOB3prH2s6BzdBtxewSweC0ibLXVxNHBJRRaJe4=";
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+, poetry-core
+, pythonOlder
+, nix-update-script
+}:
+buildPythonPackage rec {
+  pname = "keyring-pass";
+  version = "0.9.2";
+  disabled = pythonOlder "3.6";
+
+  passthru.updateScript = nix-update-script { };
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:16]:
```suggestion
```
I don't think that's needed for Python packages.

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 17]
```diff
@@ -0,0 +1,80 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, gnupg
+, keyring
+, pass
+, poetry-core
+, pythonOlder
+, nix-update-script
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:9]:
```suggestion
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    Key-Length: 1024
+    Subkey-Type: 1
+    Subkey-Length: 1024
+    Name-Real: test
+    Name-Email: test@example.com
+    Expire-Date: 1
+    EOF
+
+    # configure password store
+    pass init test@example.com
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:56]:
```suggestion
    ${lib.getExe pass} init test@example.com
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  '';
+
+  nativeBuildInputs = [
+    poetry-core
+  ];
+
+  nativeCheckInputs = [
+    keyring
+    gnupg
+    pass
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:36]:
```suggestion
```

File: pkgs/development/python-modules/keyring-pass/default.nix [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  src = fetchFromGitHub {
+    owner = "nazarewk";
+    repo = "keyring_pass";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-Sf7eDOB3prH2s6BzdBtxewSweC0ibLXVxNHBJRRaJe4=";
+  };
+
+  postPatch = ''
+    substituteInPlace keyring_pass/__init__.py \
+      --replace 'pass_binary = "pass"' 'pass_binary = "${pass}/bin/pass"'
```
Comment (dotlambda) [inline pkgs/development/python-modules/keyring-pass/default.nix:26]:
```suggestion
      --replace-fail 'pass_binary = "pass"' 'pass_binary = "${lib.getExe pass}"'
```

**Review by dotlambda — COMMENTED** [review]
I just had an idea for how to make the test fail if the `pass_binary` substitution fails or is insufficient. Feel free to leave things as they are though if you prefer.

**Comment by nazarewk** [comment]
> I just had an idea for how to make the test fail if the `pass_binary` substitution fails or is insufficient. Feel free to leave things as they are though if you prefer.

seems like a good idea

File: pkgs/development/python-modules/keyring-pass/default.nix (hunk 20 above)
Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix:26]: nice, didn't know about this one

Comment (nazarewk) [reply pkgs/development/python-modules/keyring-pass/default.nix]: I was right not to know about it, it wasn't there when I started the PR, had to rebase onto `nixos-unstable` ;

**Review by dotlambda — APPROVED** [review]
